sql/atxs/atxs.go

```diff
@@ -0,0 +1,237 @@
+	}
+
+	return id, err
+}
+
+// GetIDsByEpoch gets ATX IDs for a given epoch.
+func GetIDsByEpoch(db sql.Executor, epoch types.EpochID) (ids []types.ATXID, err error) {
+	enc := func(stmt *sql.Statement) {
+		stmt.BindInt64(1, int64(epoch))
+	}
+	dec := func(stmt *sql.Statement) bool {
+		var id types.ATXID
+		stmt.ColumnBytes(0, id[:])
+		ids = append(ids, id)
+		return true
+	}
+
+	if rows, err := db.Exec("select id from atxs where epoch = ?1;", enc, dec); err != nil {
+		return nil, fmt.Errorf("exec epoch %v: %w", epoch, err)
+	} else if rows == 0 {
+		return []types.ATXID{}, nil
+	}
+
+	return ids, nil
+}
+
+// GetBlob loads ATX as an encoded blob, ready to be sent over the wire.
+func GetBlob(db sql.Executor, id types.ATXID) (buf []byte, err error) {
+	if rows, err := db.Exec("select atx from atxs where id = ?1",
+		func(stmt *sql.Statement) {
+			stmt.BindBytes(1, id.Bytes())
+		}, func(stmt *sql.Statement) bool {
+			buf = make([]byte, stmt.ColumnLen(0))
+			stmt.ColumnBytes(0, buf)
+			return true
+		}); err != nil {
+		return nil, fmt.Errorf("get %s: %w", id, err)
+	} else if rows == 0 {
+		return nil, fmt.Errorf("%w: atx %s", sql.ErrNotFound, id)
+	}
+	return buf, nil
+}
+
+// Add adds an ATX for a given ATX ID.
+func Add(db sql.Executor, atx *types.ActivationTx, timestamp time.Time) error {
+	buf, err := codec.Encode(atx)
+	if err != nil {
+		return fmt.Errorf("encode: %w", err)
+	}
+
+	enc := func(stmt *sql.Statement) {
+		stmt.BindBytes(1, atx.ID().Bytes())
+		stmt.BindInt64(2, int64(atx.PubLayerID.Uint32()))
+		stmt.BindInt64(3, int64(atx.PubLayerID.GetEpoch()))
+		stmt.BindBytes(4, atx.NodeID.ToBytes())
+		stmt.BindBytes(5, buf)
+		stmt.BindInt64(6, timestamp.UnixNano())
+	}
+
+	_, err = db.Exec(`
+		insert into atxs (id, layer, epoch, smesher, atx, timestamp) 
+		values (?1, ?2, ?3, ?4, ?5, ?6);`, enc, nil)
+	if err != nil {
+		return fmt.Errorf("insert ATX ID %v: %w", atx.ID(), err)
+	}
+
+	return nil
+}
+
+// GetTop gets a top ATX (positioning ATX candidate) for a given ATX ID.
+// The top ATX is the one with the highest layer ID.
+func GetTop(db sql.Executor) (id types.ATXID, err error) {
+	dec := func(stmt *sql.Statement) bool {
+		stmt.ColumnBytes(0, id[:])
+		return true
+	}
+
+	if rows, err := db.Exec("select atx_id from atx_top;", nil, dec); err != nil {
+		return types.ATXID{}, fmt.Errorf("exec: %w", err)
+	} else if rows == 0 {
+		return types.ATXID{}, fmt.Errorf("exec: %w", sql.ErrNotFound)
+	}
+
+	return id, err
+}
+
+// UpdateTopIfNeeded updates top ATX if needed
+func UpdateTopIfNeeded(db sql.Executor, atx *types.ActivationTx) error {
+	enc := func(stmt *sql.Statement) {
+		stmt.BindBytes(1, atx.ID().Bytes())
+		stmt.BindInt64(2, int64(atx.PubLayerID.Uint32()))
+	}
+
+	_, err := db.Exec(`
+		insert into atx_top (id, atx_id, layer) 
+		values (1, ?1, ?2)
+		on conflict (id) do
+		update set 
+			layer = case
+				when excluded.layer > layer
+				then excluded.layer
+				else layer
+			end,
```

Comment:
i don't understand `excluded` variable, how does it work? i thought that this statement should be smth like:

```
layer = case
when layer > ?1 
then layer else ?1 
end
```

Reply:
The result is in fact the same as `?1`, it's a reference to the row that is being inserted. I can change it to `?1` for consistency with other code if you want
https://www.sqlite.org/lang_UPSERT.html
```
Column names in the expressions of a DO UPDATE refer to the original unchanged value of the column, before the attempted INSERT. To use the value that would have been inserted had the constraint not failed, add the special "excluded." table qualifier to the column name.
```

Reply:
thanks, i think consistency is important. but we can also use excluded everywhere as it is mentioned in sqlite docs

Reply:
explain with excluded
```
+----+-----------+-----+--+--+------------------------------------------------------------------+--+-------+
|addr|opcode     |p1   |p2|p3|p4                                                                |p5|comment|
+----+-----------+-----+--+--+------------------------------------------------------------------+--+-------+
|0   |Init       |0    |43|0 |NULL                                                              |0 |NULL   |
|1   |OpenWrite  |0    |24|0 |3                                                                 |0 |NULL   |
|2   |OpenWrite  |1    |25|0 |k(2,,)                                                            |0 |NULL   |
|3   |Integer    |1    |2 |0 |NULL                                                              |0 |NULL   |
|4   |String8    |0    |3 |0 |0x07FACD1586DD5FD20A1759F35FDDDF46802E6F47446187CC12E9F263AF7A1281|0 |NULL   |
|5   |Cast       |3    |66|0 |NULL                                                              |0 |NULL   |
|6   |Integer    |12345|4 |0 |NULL                                                              |0 |NULL   |
|7   |NewRowid   |0    |1 |0 |NULL                                                              |0 |NULL   |
|8   |HaltIfNull |1299 |2 |4 |atx_top.layer                                                     |1 |NULL   |
|9   |Affinity   |2    |3 |0 |DBD                                                               |0 |NULL   |
|10  |Eq         |10   |12|2 |BINARY-8                                                          |84|NULL   |
|11  |Halt       |275  |2 |0 |id = 1                                                            |3 |NULL   |
|12  |Goto       |0    |14|0 |NULL                                                              |0 |NULL   |
|13  |Goto       |0    |39|0 |NULL                                                              |0 |NULL   |
|14  |SCopy      |2    |6 |0 |NULL                                                              |0 |NULL   |
|15  |IntCopy    |1    |7 |0 |NULL                                                              |0 |NULL   |
|16  |MakeRecord |6    |2 |5 |NULL                                                              |0 |NULL   |
|17  |NoConflict |1    |13|6 |1                                                                 |0 |NULL   |
|18  |IdxRowid   |1    |11|0 |NULL                                                              |0 |NULL   |
|19  |SeekRowid  |0    |0 |11|NULL                                                              |0 |NULL   |
|20  |Null       |0    |12|13|NULL                                                              |0 |NULL   |
|21  |Noop       |2    |0 |12|NULL                                                              |0 |NULL   |
|22  |Rowid      |0    |13|0 |NULL                                                              |0 |NULL   |
|23  |Column     |0    |0 |14|NULL                                                              |0 |NULL   |
|24  |Column     |0    |2 |11|NULL                                                              |0 |NULL   |
|25  |Le         |11   |28|4 |BINARY-8                                                          |84|NULL   |
|26  |SCopy      |3    |15|0 |NULL                                                              |0 |NULL   |
|27  |Goto       |0    |29|0 |NULL                                                              |0 |NULL   |
|28  |Column     |0    |1 |15|NULL                                                              |0 |NULL   |
|29  |Column     |0    |2 |11|NULL                                                              |0 |NULL   |
|30  |Le         |11   |33|4 |BINARY-8                                                          |84|NULL   |
|31  |SCopy      |4    |16|0 |NULL                                                              |0 |NULL   |
|32  |Goto       |0    |34|0 |NULL                                                              |0 |NULL   |
|33  |Column     |0    |2 |16|NULL                                                              |0 |NULL   |
|34  |HaltIfNull |1299 |2 |16|atx_top.layer                                                     |1 |NULL   |
|35  |MakeRecord |14   |3 |12|DBD                                                               |0 |NULL   |
|36  |Insert     |0    |12|13|atx_top                                                           |5 |NULL   |
|37  |Goto       |0    |42|0 |NULL                                                              |0 |NULL   |
|38  |Goto       |0    |13|0 |NULL                                                              |0 |NULL   |
|39  |MakeRecord |2    |3 |8 |NULL                                                              |0 |NULL   |
|40  |IdxInsert  |1    |5 |6 |2                                                                 |16|NULL   |
|41  |Insert     |0    |8 |1 |atx_top                                                           |57|NULL   |
|42  |Halt       |0    |0 |0 |NULL                                                              |0 |NULL   |
|43  |Transaction|0    |1 |17|0                                                                 |1 |NULL   |
|44  |Integer    |1    |10|0 |NULL                                                              |0 |NULL   |
|45  |Goto       |0    |1 |0 |NULL                                                              |0 |NULL   |
+----+-----------+-----+--+--+------------------------------------------------------------------+--+-------+
```

explain without excluded 
```
+----+-----------+-----+--+--+------------------------------------------------------------------+--+-------+
|addr|opcode     |p1   |p2|p3|p4                                                                |p5|comment|
+----+-----------+-----+--+--+------------------------------------------------------------------+--+-------+
|0   |Init       |0    |44|0 |NULL                                                              |0 |NULL   |
|1   |OpenWrite  |0    |24|0 |3                                                                 |0 |NULL   |
|2   |OpenWrite  |1    |25|0 |k(2,,)                                                            |0 |NULL   |
|3   |Integer    |1    |2 |0 |NULL                                                              |0 |NULL   |
|4   |String8    |0    |3 |0 |0x07FACD1586DD5FD20A1759F35FDDDF46802E6F47446187CC12E9F263AF7A1281|0 |NULL   |
|5   |Cast       |3    |66|0 |NULL                                                              |0 |NULL   |
|6   |Integer    |12345|4 |0 |NULL                                                              |0 |NULL   |
|7   |NewRowid   |0    |1 |0 |NULL                                                              |0 |NULL   |
|8   |HaltIfNull |1299 |2 |4 |atx_top.layer                                                     |1 |NULL   |
|9   |Affinity   |2    |3 |0 |DBD                                                               |0 |NULL   |
|10  |Eq         |10   |12|2 |BINARY-8                                                          |84|NULL   |
|11  |Halt       |275  |2 |0 |id = 1                                                            |3 |NULL   |
|12  |Goto       |0    |14|0 |NULL                                                              |0 |NULL   |
|13  |Goto       |0    |40|0 |NULL                                                              |0 |NULL   |
|14  |SCopy      |2    |6 |0 |NULL                                                              |0 |NULL   |
|15  |IntCopy    |1    |7 |0 |NULL                                                              |0 |NULL   |
|16  |MakeRecord |6    |2 |5 |NULL                                                              |0 |NULL   |
|17  |NoConflict |1    |13|6 |1                                                                 |0 |NULL   |
|18  |IdxRowid   |1    |11|0 |NULL                                                              |0 |NULL   |
|19  |SeekRowid  |0    |0 |11|NULL                                                              |0 |NULL   |
|20  |Null       |0    |12|13|NULL                                                              |0 |NULL   |
|21  |Noop       |2    |0 |12|NULL                                                              |0 |NULL   |
|22  |Rowid      |0    |13|0 |NULL                                                              |0 |NULL   |
|23  |Column     |0    |0 |14|NULL                                                              |0 |NULL   |
|24  |Column     |0    |2 |11|NULL                                                              |0 |NULL   |
|25  |Le         |11   |29|17|BINARY-8                                                          |84|NULL   |
|26  |String8    |0    |15|0 |0x07FACD1586DD5FD20A1759F35FDDDF46802E6F47446187CC12E9F263AF7A1281|0 |NULL   |
|27  |Cast       |15   |66|0 |NULL                                                              |0 |NULL   |
|28  |Goto       |0    |30|0 |NULL                                                              |0 |NULL   |
|29  |Column     |0    |1 |15|NULL                                                              |0 |NULL   |
|30  |Column     |0    |2 |11|NULL                                                              |0 |NULL   |
|31  |Le         |11   |34|17|BINARY-8                                                          |84|NULL   |
|32  |Integer    |12345|16|0 |NULL                                                              |0 |NULL   |
|33  |Goto       |0    |35|0 |NULL                                                              |0 |NULL   |
|34  |Column     |0    |2 |16|NULL                                                              |0 |NULL   |
|35  |HaltIfNull |1299 |2 |16|atx_top.layer                                                     |1 |NULL   |
|36  |MakeRecord |14   |3 |12|DBD                                                               |0 |NULL   |
|37  |Insert     |0    |12|13|atx_top                                                           |5 |NULL   |
|38  |Goto       |0    |43|0 |NULL                                                              |0 |NULL   |
|39  |Goto       |0    |13|0 |NULL                                                              |0 |NULL   |
|40  |MakeRecord |2    |3 |8 |NULL                                                              |0 |NULL   |
|41  |IdxInsert  |1    |5 |6 |2                                                                 |16|NULL   |
|42  |Insert     |0    |8 |1 |atx_top                                                           |57|NULL   |
|43  |Halt       |0    |0 |0 |NULL                                                              |0 |NULL   |
|44  |Transaction|0    |1 |17|0                                                                 |1 |NULL   |
|45  |Integer    |1    |10|0 |NULL                                                              |0 |NULL   |
|46  |Integer    |12345|17|0 |NULL                                                              |0 |NULL   |
|47  |Goto       |0    |1 |0 |NULL                                                              |0 |NULL   |
+----+-----------+-----+--+--+------------------------------------------------------------------+--+-------+
```